lls/cache.c

```diff
@@ -0,0 +1,428 @@
+	const void *ip_key, const struct ether_addr *ha, uint8_t port_id,
+	enum lls_cache_op op)
+{
+	struct lls_cache_entry *lls_entry = NULL;
+	struct lls_cache_req *req;
+	int ret;
+
+	if (unlikely(cache == NULL)) {
+		RTE_LOG(ERR, GATEKEEPER, "lls: %s: %s cache not instantiated, can't perform op %d\n",
+			__func__, cache->name, op);
+		return -EINVAL;
+	}
+
+	req = mb_alloc_entry(mb);
+	if (unlikely(req == NULL))
+		return -1;
+
+	if (op == LLS_CACHE_DEC)
+		goto send;
+
+	/*
+	 * Allocate a new entry. If one already exists for @ip_key,
+	 * it will eventually be thrown away.
+	 */
+	ret = rte_mempool_get(cache->entry_pool, (void **)&lls_entry);
+	if (unlikely(ret == -ENOENT)) {
+		mb_free_entry(mb, req);
+		RTE_LOG(ERR, MEMPOOL,
+			"No memory for new entry, %s entry not added\n",
+			cache->name);
+		return ret;
+	}
+
+	switch (op) {
+	case LLS_CACHE_INC:
+		lls_entry->stale = true;
+		lls_entry->valid = false;
+		lls_entry->refcnt = 1;
+		break;
+	case LLS_CACHE_ADD:
+		ether_addr_copy(ha, &lls_entry->ha);
+		lls_entry->port_id = port_id;
+		lls_entry->stale = false;
+		lls_entry->valid = true;
+		lls_entry->refcnt = 0;
+		break;
+	default:
+		rte_mempool_put(cache->entry_pool, lls_entry);
+		mb_free_entry(mb, req);
+		RTE_LOG(ERR, GATEKEEPER, "%s cache op (%d) invalid\n",
+			cache->name, op);
+		return -1;
+	}
+
+	lls_entry->ts = time(NULL);
+
+send:
+	req->entry = lls_entry;
+	req->op = op;
+	req->cache = cache;
+	memcpy(req->ip_key, ip_key, cache->key_len);
+
+	ret = mb_send_entry(mb, req);
+	if (ret != 0 && op != LLS_CACHE_DEC)
+		rte_mempool_put(cache->entry_pool, lls_entry);
+	return ret;
+}
+
+void
+lls_cache_scan(struct lls_cache *cache, struct gatekeeper_if *front,
+	struct gatekeeper_if *back)
+{
+	uint32_t iter = 0;
+	int32_t index;
+	const void *ip_key;
+	void *data;
+	time_t now = time(NULL);
+
+	index = rte_hash_iterate(cache->hash, &ip_key, &data, &iter);
+	while (index >= 0) {
+		struct lls_cache_entry *lls_entry = cache->data[index];
+		uint32_t timeout;
+
+		/*
+		 * If an entry is already stale or invalid, continue
+		 * to try to resolve it while there's interest.
+		 */
+		if (!lls_entry->valid || lls_entry->stale) {
+			if (lls_entry->refcnt > 0) {
+				lls_entry->valid = false;
+				/* Broadcast request on both ports. */
+				cache->xmit_req(front->id, ip_key, NULL);
+				cache->xmit_req(back->id, ip_key, NULL);
+			} else
+				__lls_cache_del(cache, ip_key);
+			goto next;
+		}
+
+		if (lls_entry->port_id == front->id)
+			timeout = cache->front_timeout_sec;
+		else if (back != NULL && lls_entry->port_id == back->id)
+			timeout = cache->back_timeout_sec;
+		else {
+			RTE_LOG(ERR, GATEKEEPER, "lls: %s cache has an entry associated with neither the front nor back port (%hhu)\n",
+				cache->name, lls_entry->port_id);
+			goto next;
+		}
+
+		if (now - lls_entry->ts > timeout) {
+			lls_entry->stale = true;
+			if (lls_entry->refcnt > 0) {
+				/* Unicast request. */
+				cache->xmit_req(lls_entry->port_id, ip_key,
```

Comment:
If the IP address now maps to another MAC address, the unicast should not get a response. When the scan happens again, if a response has still not been heard the broadcast will happen.

I think it would also be OK to remove the unicast and just always do broadcast, since entries will only turn stale every two hours or so.